libs/pdi-scanner/src/rusb_async/mod.rs

```diff
@@ -0,0 +1,324 @@
+// File originally copied from https://github.com/a1ien/rusb/commit/4d955e48394e2a5c116fd4e3c9408c88536ed894,
+// with the license:
+//
+//     Copyright (c) 2015 David Cuddeback
+//                   2019 Ilya Averyanov
+//
+//     Permission is hereby granted, free of charge, to any person obtaining
+//     a copy of this software and associated documentation files (the
+//     "Software"), to deal in the Software without restriction, including
+//     without limitation the rights to use, copy, modify, merge, publish,
+//     distribute, sublicense, and/or sell copies of the Software, and to
+//     permit persons to whom the Software is furnished to do so, subject to
+//     the following conditions:
+//
+//     The above copyright notice and this permission notice shall be
+//     included in all copies or substantial portions of the Software.
+//
+//     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+//     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+//     MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+//     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+//     LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+//     OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+//     WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// Modifications have been made to better support the use case of this project.
+
+use rusb::ffi;
+
+use std::convert::TryInto;
+use std::ptr::NonNull;
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+mod error;
+mod pool;
+
+pub use error::{Error, Result};
+pub use pool::TransferPool;
+
+struct Transfer {
+    ptr: NonNull<ffi::libusb_transfer>,
+    buffer: Vec<u8>,
+}
+
+impl Transfer {
+    // Invariant: Caller must ensure `device` outlives this transfer
+    #[allow(unused)]
```

Comment:
This was added to make `clippy` happy for each of the methods. In reality, we do use `bulk` but not the others so we could delete the others and remove this. I opted to keep it closer to the original instead.